opm/simulators/linalg/OwningTwoLevelPreconditioner.hpp

```diff
@@ -69,51 +86,61 @@ class OwningTwoLevelPreconditioner : public Dune::PreconditionerWithUpdate<Vecto
                            finesmoother_,
                            levelTransferPolicy_,
                            coarseSolverPolicy_,
-                           transpose ? 1 : 0,
-                           transpose ? 0 : 1)
+                           prm.get<int>("pre_smooth", transpose? 1 : 0),
+                           prm.get<int>("post_smooth", transpose? 0 : 1))
         , prm_(prm)
     {
-        if (prm.get<int>("verbosity") > 10) {
-            std::ofstream outfile(prm.get<std::string>("weights_filename"));
+        if (prm.get<int>("verbosity", 0) > 10) {
+            std::string filename = prm.get<std::string>("weights_filename", "impes_weights.txt");
+            std::ofstream outfile(filename);
             if (!outfile) {
-                throw std::runtime_error("Could not write weights");
+                OPM_THROW(std::ofstream::failure, "Could not write weights to file " << filename << ".");
             }
             Dune::writeMatrixMarket(weights_, outfile);
         }
     }
 
-    OwningTwoLevelPreconditioner(const OperatorType& linearoperator, const pt& prm, const Communication& comm)
+    OwningTwoLevelPreconditioner(const OperatorType& linearoperator, const pt& prm,
+                                 const std::function<VectorType()> weightsCalculator, const Communication& comm)
         : linear_operator_(linearoperator)
-        , finesmoother_(PrecFactory::create(linearoperator, prm.get_child("finesmoother"), comm))
+        , finesmoother_(PrecFactory::create(linearoperator,
+                                            prm.get_child_optional("finesmoother")?
+                                            prm.get_child("finesmoother"): pt(), comm))
         , comm_(&comm)
-        , weights_(Opm::Amg::getQuasiImpesWeights<MatrixType, VectorType>(
-              linearoperator.getmat(), prm.get<int>("pressure_var_index"), transpose))
-        , levelTransferPolicy_(*comm_, weights_, prm.get<int>("pressure_var_index"))
-        , coarseSolverPolicy_(prm.get_child("coarsesolver"))
+        , weightsCalculator_(weightsCalculator)
+        , weights_(weightsCalculator())
+        , levelTransferPolicy_(*comm_, weights_, prm.get<int>("pressure_var_index", 1))
+        , coarseSolverPolicy_(prm.get_child_optional("coarsesolver")? prm.get_child("coarsesolver") : pt())
         , twolevel_method_(linearoperator,
                            finesmoother_,
                            levelTransferPolicy_,
                            coarseSolverPolicy_,
-                           transpose ? 1 : 0,
-                           transpose ? 0 : 1)
+                           prm.get<int>("pre_smooth", transpose? 1 : 0),
+                           prm.get<int>("post_smooth", transpose? 0 : 1))
         , prm_(prm)
     {
-        if (prm.get<int>("verbosity") > 10) {
-            std::ofstream outfile(prm.get<std::string>("weights_filename"));
+        //Opm::Amg::getQuasiImpesWeights<MatrixType, VectorType>(
```

Comment:
Delete unused.

Reply:
Probably leftover from the original PR. Will remove